src/core/thread/router_table.cpp

```diff
@@ -242,9 +264,17 @@ Router *RouterTable::Allocate(void)
 Router *RouterTable::Allocate(uint8_t aRouterId)
 {
     Router *rval = nullptr;
+#if OPENTHREAD_CONFIG_REFERENCE_DEVICE_ENABLE
+    uint8_t minRouterIdAllowed = 0, maxRouterIdAllowed = Mle::kMaxRouterId;
 
+    GetAllowedRouterIdRange(&minRouterIdAllowed, &maxRouterIdAllowed);
+    VerifyOrExit(minRouterIdAllowed <= aRouterId && aRouterId <= maxRouterIdAllowed &&
+                 mActiveRouterCount < Mle::kMaxRouters && !IsAllocated(aRouterId) &&
+                 mRouterIdReuseDelay[aRouterId] == 0);
+#else
     VerifyOrExit(aRouterId <= Mle::kMaxRouterId && mActiveRouterCount < Mle::kMaxRouters && !IsAllocated(aRouterId) &&
                  mRouterIdReuseDelay[aRouterId] == 0);
+#endif
```

Comment:
I think it's better to just allow allocating any Router ID to 1) minimize code change and 2) not change Thread logic.

A Thread network will never try to allocate a Router ID out of range if it was created with a given Router ID range.